{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pathlib\n",
    "\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Parse the NCBI BLAST outputs\n",
    "\n",
    "1. We BLAST the query sequences in `sequences.fna` against the NCBI 16S reference database using the web tool\n",
    "2. The outputs are downloaded into `ncbi16s_desc.csv` and `ncbi16s_hits.csv`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path = pathlib.Path(\"../../data/processed/kchip\")\n",
    "blast_path = data_path / \"blast\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Headers for the hits and desc files\n",
    "ncbi_hits_header = \"qid,Accession,pidentity,length,mismatch,gapopen,qstart,qend,sstart,send,evalue,bitscore\".split(\",\")\n",
    "ncbi_desc_header = \"Description,Scientific Name,Max Score,Total Score,Query Cover,E value,Per. ident,Acc. Len,Accession Full\".split(\",\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "ncbi_blast_hits_file = blast_path / \"ncbi16s_hits.csv\"\n",
    "ncbi_blast_desc_file = blast_path / \"ncbi16s_desc.csv\"\n",
    "ncbi_blast_hits = pd.read_csv(ncbi_blast_hits_file, sep=\",\", names=ncbi_hits_header)\n",
    "ncbi_blast_desc = pd.read_csv(ncbi_blast_desc_file, sep=\",\", skiprows=1, names=ncbi_desc_header)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>qid</th>\n",
       "      <th>Accession</th>\n",
       "      <th>pidentity</th>\n",
       "      <th>length</th>\n",
       "      <th>mismatch</th>\n",
       "      <th>gapopen</th>\n",
       "      <th>qstart</th>\n",
       "      <th>qend</th>\n",
       "      <th>sstart</th>\n",
       "      <th>send</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bitscore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CR-75</td>\n",
       "      <td>NR_180446.1</td>\n",
       "      <td>98.587</td>\n",
       "      <td>1415</td>\n",
       "      <td>13</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>1411</td>\n",
       "      <td>52</td>\n",
       "      <td>1463</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2495</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CR-75</td>\n",
       "      <td>NR_180451.1</td>\n",
       "      <td>98.303</td>\n",
       "      <td>1414</td>\n",
       "      <td>19</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>1411</td>\n",
       "      <td>49</td>\n",
       "      <td>1460</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2473</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CR-75</td>\n",
       "      <td>NR_126208.1</td>\n",
       "      <td>98.292</td>\n",
       "      <td>1405</td>\n",
       "      <td>19</td>\n",
       "      <td>5</td>\n",
       "      <td>10</td>\n",
       "      <td>1411</td>\n",
       "      <td>1</td>\n",
       "      <td>1403</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2457</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CR-75</td>\n",
       "      <td>NR_028894.1</td>\n",
       "      <td>98.092</td>\n",
       "      <td>1415</td>\n",
       "      <td>20</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>1411</td>\n",
       "      <td>32</td>\n",
       "      <td>1443</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2457</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CR-75</td>\n",
       "      <td>NR_028687.1</td>\n",
       "      <td>98.092</td>\n",
       "      <td>1415</td>\n",
       "      <td>20</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>1411</td>\n",
       "      <td>46</td>\n",
       "      <td>1457</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2457</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>843</th>\n",
       "      <td>O-G1</td>\n",
       "      <td>NR_042356.1</td>\n",
       "      <td>97.660</td>\n",
       "      <td>1410</td>\n",
       "      <td>29</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>1406</td>\n",
       "      <td>17</td>\n",
       "      <td>1426</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2418</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>844</th>\n",
       "      <td>O-G1</td>\n",
       "      <td>NR_159317.1</td>\n",
       "      <td>98.014</td>\n",
       "      <td>1410</td>\n",
       "      <td>26</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1408</td>\n",
       "      <td>25</td>\n",
       "      <td>1434</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2451</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>845</th>\n",
       "      <td>O-G1</td>\n",
       "      <td>NR_180640.1</td>\n",
       "      <td>97.872</td>\n",
       "      <td>1410</td>\n",
       "      <td>28</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1408</td>\n",
       "      <td>52</td>\n",
       "      <td>1461</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2436</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>846</th>\n",
       "      <td>O-G1</td>\n",
       "      <td>NR_041927.1</td>\n",
       "      <td>97.943</td>\n",
       "      <td>1410</td>\n",
       "      <td>27</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1408</td>\n",
       "      <td>21</td>\n",
       "      <td>1430</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2451</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>847</th>\n",
       "      <td>O-G1</td>\n",
       "      <td>NR_126319.1</td>\n",
       "      <td>97.805</td>\n",
       "      <td>1412</td>\n",
       "      <td>27</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>1408</td>\n",
       "      <td>41</td>\n",
       "      <td>1452</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2433</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>848 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       qid    Accession  pidentity  length  mismatch  gapopen  qstart  qend  \\\n",
       "0    CR-75  NR_180446.1     98.587    1415        13        7       1  1411   \n",
       "1    CR-75  NR_180451.1     98.303    1414        19        5       1  1411   \n",
       "2    CR-75  NR_126208.1     98.292    1405        19        5      10  1411   \n",
       "3    CR-75  NR_028894.1     98.092    1415        20        6       1  1411   \n",
       "4    CR-75  NR_028687.1     98.092    1415        20        6       1  1411   \n",
       "..     ...          ...        ...     ...       ...      ...     ...   ...   \n",
       "843   O-G1  NR_042356.1     97.660    1410        29        4       1  1406   \n",
       "844   O-G1  NR_159317.1     98.014    1410        26        2       1  1408   \n",
       "845   O-G1  NR_180640.1     97.872    1410        28        2       1  1408   \n",
       "846   O-G1  NR_041927.1     97.943    1410        27        2       1  1408   \n",
       "847   O-G1  NR_126319.1     97.805    1412        27        4       1  1408   \n",
       "\n",
       "     sstart  send  evalue  bitscore  \n",
       "0        52  1463     0.0      2495  \n",
       "1        49  1460     0.0      2473  \n",
       "2         1  1403     0.0      2457  \n",
       "3        32  1443     0.0      2457  \n",
       "4        46  1457     0.0      2457  \n",
       "..      ...   ...     ...       ...  \n",
       "843      17  1426     0.0      2418  \n",
       "844      25  1434     0.0      2451  \n",
       "845      52  1461     0.0      2436  \n",
       "846      21  1430     0.0      2451  \n",
       "847      41  1452     0.0      2433  \n",
       "\n",
       "[848 rows x 12 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filter the blast hits based on E-value\n",
    "ncbi_blast_hits = ncbi_blast_hits[ncbi_blast_hits.evalue <= 1e-50]\n",
    "ncbi_blast_hits[\"qid\"] = [qid.split(\"_\")[0] for qid in ncbi_blast_hits.qid]\n",
    "ncbi_blast_hits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "qid\n",
       "CR-75                    100\n",
       "K-A10                     54\n",
       "K-A4                      54\n",
       "K-A7                      79\n",
       "K-C10                     47\n",
       "K-D4                       8\n",
       "K-D8                      71\n",
       "K-F4                      42\n",
       "K-F8                      51\n",
       "K-H10                      7\n",
       "M-B5                      58\n",
       "M-C10                     82\n",
       "M-G11                     81\n",
       "O-G1                      73\n",
       "escherichia-coli-k-12     41\n",
       "Name: Accession, dtype: int64"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Display the number of hits for each query sequence\n",
    "ncbi_blast_hits.groupby(\"qid\")[\"Accession\"].count()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We need to parse the accession id from the accession string in the description file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Description</th>\n",
       "      <th>Scientific Name</th>\n",
       "      <th>Max Score</th>\n",
       "      <th>Total Score</th>\n",
       "      <th>Query Cover</th>\n",
       "      <th>E value</th>\n",
       "      <th>Per. ident</th>\n",
       "      <th>Acc. Len</th>\n",
       "      <th>Accession Full</th>\n",
       "      <th>Accession</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Scandinavium goeteborgense strain CCUG 66741 1...</td>\n",
       "      <td>Scandinavium goeteborgense</td>\n",
       "      <td>2495</td>\n",
       "      <td>2495</td>\n",
       "      <td>100%</td>\n",
       "      <td>0.0</td>\n",
       "      <td>98.59</td>\n",
       "      <td>1542</td>\n",
       "      <td>=HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...</td>\n",
       "      <td>NR_180446.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Enterobacter quasihormaechei strain WCHEs12000...</td>\n",
       "      <td>Enterobacter quasihormaechei</td>\n",
       "      <td>2473</td>\n",
       "      <td>2473</td>\n",
       "      <td>100%</td>\n",
       "      <td>0.0</td>\n",
       "      <td>98.30</td>\n",
       "      <td>1538</td>\n",
       "      <td>=HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...</td>\n",
       "      <td>NR_180451.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Enterobacter hormaechei subsp. xiangfangensis ...</td>\n",
       "      <td>Enterobacter hormaechei subsp. xiangfangensis</td>\n",
       "      <td>2457</td>\n",
       "      <td>2457</td>\n",
       "      <td>99%</td>\n",
       "      <td>0.0</td>\n",
       "      <td>98.29</td>\n",
       "      <td>1429</td>\n",
       "      <td>=HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...</td>\n",
       "      <td>NR_126208.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Citrobacter freundii ATCC 8090 = MTCC 1658 = N...</td>\n",
       "      <td>Citrobacter freundii ATCC 8090 = MTCC 1658 = N...</td>\n",
       "      <td>2457</td>\n",
       "      <td>2457</td>\n",
       "      <td>100%</td>\n",
       "      <td>0.0</td>\n",
       "      <td>98.09</td>\n",
       "      <td>1505</td>\n",
       "      <td>=HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...</td>\n",
       "      <td>NR_028894.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Citrobacter braakii strain 167 16S ribosomal R...</td>\n",
       "      <td>Citrobacter braakii</td>\n",
       "      <td>2457</td>\n",
       "      <td>2457</td>\n",
       "      <td>100%</td>\n",
       "      <td>0.0</td>\n",
       "      <td>98.09</td>\n",
       "      <td>1530</td>\n",
       "      <td>=HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...</td>\n",
       "      <td>NR_028687.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>Raoultella terrigena strain 84 16S ribosomal R...</td>\n",
       "      <td>Raoultella terrigena</td>\n",
       "      <td>2368</td>\n",
       "      <td>2368</td>\n",
       "      <td>100%</td>\n",
       "      <td>0.0</td>\n",
       "      <td>96.96</td>\n",
       "      <td>1454</td>\n",
       "      <td>=HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...</td>\n",
       "      <td>NR_037085.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>Cedecea lapagei strain DSM 4587 16S ribosomal ...</td>\n",
       "      <td>Cedecea lapagei</td>\n",
       "      <td>2368</td>\n",
       "      <td>2368</td>\n",
       "      <td>100%</td>\n",
       "      <td>0.0</td>\n",
       "      <td>96.96</td>\n",
       "      <td>1531</td>\n",
       "      <td>=HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...</td>\n",
       "      <td>NR_126319.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>Cronobacter universalis strain E797 16S riboso...</td>\n",
       "      <td>Cronobacter universalis</td>\n",
       "      <td>2368</td>\n",
       "      <td>2368</td>\n",
       "      <td>100%</td>\n",
       "      <td>0.0</td>\n",
       "      <td>96.89</td>\n",
       "      <td>1425</td>\n",
       "      <td>=HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...</td>\n",
       "      <td>NR_115943.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>Cronobacter sakazakii strain E601 16S ribosoma...</td>\n",
       "      <td>Cronobacter sakazakii</td>\n",
       "      <td>2368</td>\n",
       "      <td>2368</td>\n",
       "      <td>100%</td>\n",
       "      <td>0.0</td>\n",
       "      <td>96.96</td>\n",
       "      <td>1424</td>\n",
       "      <td>=HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...</td>\n",
       "      <td>NR_115942.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>Cronobacter sakazakii strain ATCC 29544 16S ri...</td>\n",
       "      <td>Cronobacter sakazakii</td>\n",
       "      <td>2368</td>\n",
       "      <td>2368</td>\n",
       "      <td>100%</td>\n",
       "      <td>0.0</td>\n",
       "      <td>96.96</td>\n",
       "      <td>1493</td>\n",
       "      <td>=HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...</td>\n",
       "      <td>NR_044076.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          Description  \\\n",
       "0   Scandinavium goeteborgense strain CCUG 66741 1...   \n",
       "1   Enterobacter quasihormaechei strain WCHEs12000...   \n",
       "2   Enterobacter hormaechei subsp. xiangfangensis ...   \n",
       "3   Citrobacter freundii ATCC 8090 = MTCC 1658 = N...   \n",
       "4   Citrobacter braakii strain 167 16S ribosomal R...   \n",
       "..                                                ...   \n",
       "95  Raoultella terrigena strain 84 16S ribosomal R...   \n",
       "96  Cedecea lapagei strain DSM 4587 16S ribosomal ...   \n",
       "97  Cronobacter universalis strain E797 16S riboso...   \n",
       "98  Cronobacter sakazakii strain E601 16S ribosoma...   \n",
       "99  Cronobacter sakazakii strain ATCC 29544 16S ri...   \n",
       "\n",
       "                                      Scientific Name  Max Score  Total Score  \\\n",
       "0                          Scandinavium goeteborgense       2495         2495   \n",
       "1                        Enterobacter quasihormaechei       2473         2473   \n",
       "2       Enterobacter hormaechei subsp. xiangfangensis       2457         2457   \n",
       "3   Citrobacter freundii ATCC 8090 = MTCC 1658 = N...       2457         2457   \n",
       "4                                 Citrobacter braakii       2457         2457   \n",
       "..                                                ...        ...          ...   \n",
       "95                               Raoultella terrigena       2368         2368   \n",
       "96                                    Cedecea lapagei       2368         2368   \n",
       "97                            Cronobacter universalis       2368         2368   \n",
       "98                              Cronobacter sakazakii       2368         2368   \n",
       "99                              Cronobacter sakazakii       2368         2368   \n",
       "\n",
       "   Query Cover  E value  Per. ident  Acc. Len  \\\n",
       "0         100%      0.0       98.59      1542   \n",
       "1         100%      0.0       98.30      1538   \n",
       "2          99%      0.0       98.29      1429   \n",
       "3         100%      0.0       98.09      1505   \n",
       "4         100%      0.0       98.09      1530   \n",
       "..         ...      ...         ...       ...   \n",
       "95        100%      0.0       96.96      1454   \n",
       "96        100%      0.0       96.96      1531   \n",
       "97        100%      0.0       96.89      1425   \n",
       "98        100%      0.0       96.96      1424   \n",
       "99        100%      0.0       96.96      1493   \n",
       "\n",
       "                                       Accession Full    Accession  \n",
       "0   =HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...  NR_180446.1  \n",
       "1   =HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...  NR_180451.1  \n",
       "2   =HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...  NR_126208.1  \n",
       "3   =HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...  NR_028894.1  \n",
       "4   =HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...  NR_028687.1  \n",
       "..                                                ...          ...  \n",
       "95  =HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...  NR_037085.1  \n",
       "96  =HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...  NR_126319.1  \n",
       "97  =HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...  NR_115943.1  \n",
       "98  =HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...  NR_115942.1  \n",
       "99  =HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...  NR_044076.1  \n",
       "\n",
       "[100 rows x 10 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ncbi_blast_desc[\"Accession\"] = ncbi_blast_desc[\"Accession Full\"].apply(lambda x: x.split(\",\")[-1].strip('\")'))\n",
    "ncbi_blast_desc"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally, we join the hits and description files together"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "ncbi_blast_results = ncbi_blast_hits.merge(ncbi_blast_desc, how=\"left\", on=\"Accession\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>qid</th>\n",
       "      <th>Accession</th>\n",
       "      <th>pidentity</th>\n",
       "      <th>length</th>\n",
       "      <th>mismatch</th>\n",
       "      <th>gapopen</th>\n",
       "      <th>qstart</th>\n",
       "      <th>qend</th>\n",
       "      <th>sstart</th>\n",
       "      <th>send</th>\n",
       "      <th>...</th>\n",
       "      <th>bitscore</th>\n",
       "      <th>Description</th>\n",
       "      <th>Scientific Name</th>\n",
       "      <th>Max Score</th>\n",
       "      <th>Total Score</th>\n",
       "      <th>Query Cover</th>\n",
       "      <th>E value</th>\n",
       "      <th>Per. ident</th>\n",
       "      <th>Acc. Len</th>\n",
       "      <th>Accession Full</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CR-75</td>\n",
       "      <td>NR_180446.1</td>\n",
       "      <td>98.587</td>\n",
       "      <td>1415</td>\n",
       "      <td>13</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>1411</td>\n",
       "      <td>52</td>\n",
       "      <td>1463</td>\n",
       "      <td>...</td>\n",
       "      <td>2495</td>\n",
       "      <td>Scandinavium goeteborgense strain CCUG 66741 1...</td>\n",
       "      <td>Scandinavium goeteborgense</td>\n",
       "      <td>2495</td>\n",
       "      <td>2495</td>\n",
       "      <td>100%</td>\n",
       "      <td>0.0</td>\n",
       "      <td>98.59</td>\n",
       "      <td>1542</td>\n",
       "      <td>=HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CR-75</td>\n",
       "      <td>NR_180451.1</td>\n",
       "      <td>98.303</td>\n",
       "      <td>1414</td>\n",
       "      <td>19</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>1411</td>\n",
       "      <td>49</td>\n",
       "      <td>1460</td>\n",
       "      <td>...</td>\n",
       "      <td>2473</td>\n",
       "      <td>Enterobacter quasihormaechei strain WCHEs12000...</td>\n",
       "      <td>Enterobacter quasihormaechei</td>\n",
       "      <td>2473</td>\n",
       "      <td>2473</td>\n",
       "      <td>100%</td>\n",
       "      <td>0.0</td>\n",
       "      <td>98.30</td>\n",
       "      <td>1538</td>\n",
       "      <td>=HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CR-75</td>\n",
       "      <td>NR_126208.1</td>\n",
       "      <td>98.292</td>\n",
       "      <td>1405</td>\n",
       "      <td>19</td>\n",
       "      <td>5</td>\n",
       "      <td>10</td>\n",
       "      <td>1411</td>\n",
       "      <td>1</td>\n",
       "      <td>1403</td>\n",
       "      <td>...</td>\n",
       "      <td>2457</td>\n",
       "      <td>Enterobacter hormaechei subsp. xiangfangensis ...</td>\n",
       "      <td>Enterobacter hormaechei subsp. xiangfangensis</td>\n",
       "      <td>2457</td>\n",
       "      <td>2457</td>\n",
       "      <td>99%</td>\n",
       "      <td>0.0</td>\n",
       "      <td>98.29</td>\n",
       "      <td>1429</td>\n",
       "      <td>=HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CR-75</td>\n",
       "      <td>NR_028894.1</td>\n",
       "      <td>98.092</td>\n",
       "      <td>1415</td>\n",
       "      <td>20</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>1411</td>\n",
       "      <td>32</td>\n",
       "      <td>1443</td>\n",
       "      <td>...</td>\n",
       "      <td>2457</td>\n",
       "      <td>Citrobacter freundii ATCC 8090 = MTCC 1658 = N...</td>\n",
       "      <td>Citrobacter freundii ATCC 8090 = MTCC 1658 = N...</td>\n",
       "      <td>2457</td>\n",
       "      <td>2457</td>\n",
       "      <td>100%</td>\n",
       "      <td>0.0</td>\n",
       "      <td>98.09</td>\n",
       "      <td>1505</td>\n",
       "      <td>=HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CR-75</td>\n",
       "      <td>NR_028687.1</td>\n",
       "      <td>98.092</td>\n",
       "      <td>1415</td>\n",
       "      <td>20</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>1411</td>\n",
       "      <td>46</td>\n",
       "      <td>1457</td>\n",
       "      <td>...</td>\n",
       "      <td>2457</td>\n",
       "      <td>Citrobacter braakii strain 167 16S ribosomal R...</td>\n",
       "      <td>Citrobacter braakii</td>\n",
       "      <td>2457</td>\n",
       "      <td>2457</td>\n",
       "      <td>100%</td>\n",
       "      <td>0.0</td>\n",
       "      <td>98.09</td>\n",
       "      <td>1530</td>\n",
       "      <td>=HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>843</th>\n",
       "      <td>O-G1</td>\n",
       "      <td>NR_042356.1</td>\n",
       "      <td>97.660</td>\n",
       "      <td>1410</td>\n",
       "      <td>29</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>1406</td>\n",
       "      <td>17</td>\n",
       "      <td>1426</td>\n",
       "      <td>...</td>\n",
       "      <td>2418</td>\n",
       "      <td>Serratia ureilytica strain NiVa 51 16S ribosom...</td>\n",
       "      <td>Serratia ureilytica</td>\n",
       "      <td>2374</td>\n",
       "      <td>2374</td>\n",
       "      <td>99%</td>\n",
       "      <td>0.0</td>\n",
       "      <td>97.10</td>\n",
       "      <td>1479</td>\n",
       "      <td>=HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>844</th>\n",
       "      <td>O-G1</td>\n",
       "      <td>NR_159317.1</td>\n",
       "      <td>98.014</td>\n",
       "      <td>1410</td>\n",
       "      <td>26</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1408</td>\n",
       "      <td>25</td>\n",
       "      <td>1434</td>\n",
       "      <td>...</td>\n",
       "      <td>2451</td>\n",
       "      <td>Klebsiella grimontii strain SB73 16S ribosomal...</td>\n",
       "      <td>Klebsiella grimontii</td>\n",
       "      <td>2372</td>\n",
       "      <td>2372</td>\n",
       "      <td>100%</td>\n",
       "      <td>0.0</td>\n",
       "      <td>96.96</td>\n",
       "      <td>1454</td>\n",
       "      <td>=HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>845</th>\n",
       "      <td>O-G1</td>\n",
       "      <td>NR_180640.1</td>\n",
       "      <td>97.872</td>\n",
       "      <td>1410</td>\n",
       "      <td>28</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1408</td>\n",
       "      <td>52</td>\n",
       "      <td>1461</td>\n",
       "      <td>...</td>\n",
       "      <td>2436</td>\n",
       "      <td>Klebsiella pasteurii strain SPARK836C1 16S rib...</td>\n",
       "      <td>Klebsiella pasteurii</td>\n",
       "      <td>2372</td>\n",
       "      <td>2372</td>\n",
       "      <td>100%</td>\n",
       "      <td>0.0</td>\n",
       "      <td>97.03</td>\n",
       "      <td>1540</td>\n",
       "      <td>=HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>846</th>\n",
       "      <td>O-G1</td>\n",
       "      <td>NR_041927.1</td>\n",
       "      <td>97.943</td>\n",
       "      <td>1410</td>\n",
       "      <td>27</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1408</td>\n",
       "      <td>21</td>\n",
       "      <td>1430</td>\n",
       "      <td>...</td>\n",
       "      <td>2451</td>\n",
       "      <td>Pseudescherichia vulneris strain ATCC 33821 16...</td>\n",
       "      <td>Pseudescherichia vulneris</td>\n",
       "      <td>2370</td>\n",
       "      <td>2370</td>\n",
       "      <td>100%</td>\n",
       "      <td>0.0</td>\n",
       "      <td>96.89</td>\n",
       "      <td>1468</td>\n",
       "      <td>=HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>847</th>\n",
       "      <td>O-G1</td>\n",
       "      <td>NR_126319.1</td>\n",
       "      <td>97.805</td>\n",
       "      <td>1412</td>\n",
       "      <td>27</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>1408</td>\n",
       "      <td>41</td>\n",
       "      <td>1452</td>\n",
       "      <td>...</td>\n",
       "      <td>2433</td>\n",
       "      <td>Cedecea lapagei strain DSM 4587 16S ribosomal ...</td>\n",
       "      <td>Cedecea lapagei</td>\n",
       "      <td>2368</td>\n",
       "      <td>2368</td>\n",
       "      <td>100%</td>\n",
       "      <td>0.0</td>\n",
       "      <td>96.96</td>\n",
       "      <td>1531</td>\n",
       "      <td>=HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>848 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       qid    Accession  pidentity  length  mismatch  gapopen  qstart  qend  \\\n",
       "0    CR-75  NR_180446.1     98.587    1415        13        7       1  1411   \n",
       "1    CR-75  NR_180451.1     98.303    1414        19        5       1  1411   \n",
       "2    CR-75  NR_126208.1     98.292    1405        19        5      10  1411   \n",
       "3    CR-75  NR_028894.1     98.092    1415        20        6       1  1411   \n",
       "4    CR-75  NR_028687.1     98.092    1415        20        6       1  1411   \n",
       "..     ...          ...        ...     ...       ...      ...     ...   ...   \n",
       "843   O-G1  NR_042356.1     97.660    1410        29        4       1  1406   \n",
       "844   O-G1  NR_159317.1     98.014    1410        26        2       1  1408   \n",
       "845   O-G1  NR_180640.1     97.872    1410        28        2       1  1408   \n",
       "846   O-G1  NR_041927.1     97.943    1410        27        2       1  1408   \n",
       "847   O-G1  NR_126319.1     97.805    1412        27        4       1  1408   \n",
       "\n",
       "     sstart  send  ...  bitscore  \\\n",
       "0        52  1463  ...      2495   \n",
       "1        49  1460  ...      2473   \n",
       "2         1  1403  ...      2457   \n",
       "3        32  1443  ...      2457   \n",
       "4        46  1457  ...      2457   \n",
       "..      ...   ...  ...       ...   \n",
       "843      17  1426  ...      2418   \n",
       "844      25  1434  ...      2451   \n",
       "845      52  1461  ...      2436   \n",
       "846      21  1430  ...      2451   \n",
       "847      41  1452  ...      2433   \n",
       "\n",
       "                                           Description  \\\n",
       "0    Scandinavium goeteborgense strain CCUG 66741 1...   \n",
       "1    Enterobacter quasihormaechei strain WCHEs12000...   \n",
       "2    Enterobacter hormaechei subsp. xiangfangensis ...   \n",
       "3    Citrobacter freundii ATCC 8090 = MTCC 1658 = N...   \n",
       "4    Citrobacter braakii strain 167 16S ribosomal R...   \n",
       "..                                                 ...   \n",
       "843  Serratia ureilytica strain NiVa 51 16S ribosom...   \n",
       "844  Klebsiella grimontii strain SB73 16S ribosomal...   \n",
       "845  Klebsiella pasteurii strain SPARK836C1 16S rib...   \n",
       "846  Pseudescherichia vulneris strain ATCC 33821 16...   \n",
       "847  Cedecea lapagei strain DSM 4587 16S ribosomal ...   \n",
       "\n",
       "                                       Scientific Name Max Score  Total Score  \\\n",
       "0                           Scandinavium goeteborgense      2495         2495   \n",
       "1                         Enterobacter quasihormaechei      2473         2473   \n",
       "2        Enterobacter hormaechei subsp. xiangfangensis      2457         2457   \n",
       "3    Citrobacter freundii ATCC 8090 = MTCC 1658 = N...      2457         2457   \n",
       "4                                  Citrobacter braakii      2457         2457   \n",
       "..                                                 ...       ...          ...   \n",
       "843                                Serratia ureilytica      2374         2374   \n",
       "844                               Klebsiella grimontii      2372         2372   \n",
       "845                               Klebsiella pasteurii      2372         2372   \n",
       "846                          Pseudescherichia vulneris      2370         2370   \n",
       "847                                    Cedecea lapagei      2368         2368   \n",
       "\n",
       "     Query Cover E value  Per. ident  Acc. Len  \\\n",
       "0           100%     0.0       98.59      1542   \n",
       "1           100%     0.0       98.30      1538   \n",
       "2            99%     0.0       98.29      1429   \n",
       "3           100%     0.0       98.09      1505   \n",
       "4           100%     0.0       98.09      1530   \n",
       "..           ...     ...         ...       ...   \n",
       "843          99%     0.0       97.10      1479   \n",
       "844         100%     0.0       96.96      1454   \n",
       "845         100%     0.0       97.03      1540   \n",
       "846         100%     0.0       96.89      1468   \n",
       "847         100%     0.0       96.96      1531   \n",
       "\n",
       "                                        Accession Full  \n",
       "0    =HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...  \n",
       "1    =HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...  \n",
       "2    =HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...  \n",
       "3    =HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...  \n",
       "4    =HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...  \n",
       "..                                                 ...  \n",
       "843  =HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...  \n",
       "844  =HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...  \n",
       "845  =HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...  \n",
       "846  =HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...  \n",
       "847  =HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...  \n",
       "\n",
       "[848 rows x 21 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ncbi_blast_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "top5_blasthits_per_qid = ncbi_blast_results.groupby(\"qid\", as_index=False).apply(lambda grp: grp.nlargest(5, \"bitscore\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>qid</th>\n",
       "      <th>Accession</th>\n",
       "      <th>Scientific Name</th>\n",
       "      <th>pidentity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">0</th>\n",
       "      <th>0</th>\n",
       "      <td>CR-75</td>\n",
       "      <td>NR_180446.1</td>\n",
       "      <td>Scandinavium goeteborgense</td>\n",
       "      <td>98.587</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CR-75</td>\n",
       "      <td>NR_180451.1</td>\n",
       "      <td>Enterobacter quasihormaechei</td>\n",
       "      <td>98.303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CR-75</td>\n",
       "      <td>NR_126208.1</td>\n",
       "      <td>Enterobacter hormaechei subsp. xiangfangensis</td>\n",
       "      <td>98.292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CR-75</td>\n",
       "      <td>NR_028894.1</td>\n",
       "      <td>Citrobacter freundii ATCC 8090 = MTCC 1658 = N...</td>\n",
       "      <td>98.092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CR-75</td>\n",
       "      <td>NR_028687.1</td>\n",
       "      <td>Citrobacter braakii</td>\n",
       "      <td>98.092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">1</th>\n",
       "      <th>173</th>\n",
       "      <td>K-A10</td>\n",
       "      <td>NR_113614.1</td>\n",
       "      <td>Klebsiella aerogenes</td>\n",
       "      <td>98.935</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>176</th>\n",
       "      <td>K-A10</td>\n",
       "      <td>NR_102493.2</td>\n",
       "      <td>Klebsiella aerogenes KCTC 2190</td>\n",
       "      <td>98.935</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179</th>\n",
       "      <td>K-A10</td>\n",
       "      <td>NR_024643.1</td>\n",
       "      <td>Klebsiella aerogenes</td>\n",
       "      <td>98.933</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>177</th>\n",
       "      <td>K-A10</td>\n",
       "      <td>NR_114737.1</td>\n",
       "      <td>Klebsiella aerogenes</td>\n",
       "      <td>98.793</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>K-A10</td>\n",
       "      <td>NR_184601.1</td>\n",
       "      <td>Huaxiibacter chinensis</td>\n",
       "      <td>98.651</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">2</th>\n",
       "      <th>239</th>\n",
       "      <td>K-A4</td>\n",
       "      <td>NR_178843.1</td>\n",
       "      <td>Pantoea endophytica</td>\n",
       "      <td>95.907</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>218</th>\n",
       "      <td>K-A4</td>\n",
       "      <td>NR_104933.1</td>\n",
       "      <td>Leclercia adecarboxylata</td>\n",
       "      <td>95.760</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>222</th>\n",
       "      <td>K-A4</td>\n",
       "      <td>NR_179946.1</td>\n",
       "      <td>Enterobacter sichuanensis</td>\n",
       "      <td>95.760</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>219</th>\n",
       "      <td>K-A4</td>\n",
       "      <td>NR_117405.1</td>\n",
       "      <td>Leclercia adecarboxylata ATCC 23216 = NBRC 102595</td>\n",
       "      <td>95.689</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>228</th>\n",
       "      <td>K-A4</td>\n",
       "      <td>NR_148649.1</td>\n",
       "      <td>Enterobacter bugandensis</td>\n",
       "      <td>95.754</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">3</th>\n",
       "      <th>249</th>\n",
       "      <td>K-A7</td>\n",
       "      <td>NR_180446.1</td>\n",
       "      <td>Scandinavium goeteborgense</td>\n",
       "      <td>98.652</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254</th>\n",
       "      <td>K-A7</td>\n",
       "      <td>NR_180450.1</td>\n",
       "      <td>Enterobacter wuhouensis</td>\n",
       "      <td>97.939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>297</th>\n",
       "      <td>K-A7</td>\n",
       "      <td>NR_113614.1</td>\n",
       "      <td>Klebsiella aerogenes</td>\n",
       "      <td>97.868</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>300</th>\n",
       "      <td>K-A7</td>\n",
       "      <td>NR_102493.2</td>\n",
       "      <td>Klebsiella aerogenes KCTC 2190</td>\n",
       "      <td>97.868</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>301</th>\n",
       "      <td>K-A7</td>\n",
       "      <td>NR_114737.1</td>\n",
       "      <td>Klebsiella aerogenes</td>\n",
       "      <td>97.797</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         qid    Accession                                    Scientific Name  \\\n",
       "0 0    CR-75  NR_180446.1                         Scandinavium goeteborgense   \n",
       "  1    CR-75  NR_180451.1                       Enterobacter quasihormaechei   \n",
       "  2    CR-75  NR_126208.1      Enterobacter hormaechei subsp. xiangfangensis   \n",
       "  3    CR-75  NR_028894.1  Citrobacter freundii ATCC 8090 = MTCC 1658 = N...   \n",
       "  4    CR-75  NR_028687.1                                Citrobacter braakii   \n",
       "1 173  K-A10  NR_113614.1                               Klebsiella aerogenes   \n",
       "  176  K-A10  NR_102493.2                     Klebsiella aerogenes KCTC 2190   \n",
       "  179  K-A10  NR_024643.1                               Klebsiella aerogenes   \n",
       "  177  K-A10  NR_114737.1                               Klebsiella aerogenes   \n",
       "  157  K-A10  NR_184601.1                             Huaxiibacter chinensis   \n",
       "2 239   K-A4  NR_178843.1                                Pantoea endophytica   \n",
       "  218   K-A4  NR_104933.1                           Leclercia adecarboxylata   \n",
       "  222   K-A4  NR_179946.1                          Enterobacter sichuanensis   \n",
       "  219   K-A4  NR_117405.1  Leclercia adecarboxylata ATCC 23216 = NBRC 102595   \n",
       "  228   K-A4  NR_148649.1                           Enterobacter bugandensis   \n",
       "3 249   K-A7  NR_180446.1                         Scandinavium goeteborgense   \n",
       "  254   K-A7  NR_180450.1                            Enterobacter wuhouensis   \n",
       "  297   K-A7  NR_113614.1                               Klebsiella aerogenes   \n",
       "  300   K-A7  NR_102493.2                     Klebsiella aerogenes KCTC 2190   \n",
       "  301   K-A7  NR_114737.1                               Klebsiella aerogenes   \n",
       "\n",
       "       pidentity  \n",
       "0 0       98.587  \n",
       "  1       98.303  \n",
       "  2       98.292  \n",
       "  3       98.092  \n",
       "  4       98.092  \n",
       "1 173     98.935  \n",
       "  176     98.935  \n",
       "  179     98.933  \n",
       "  177     98.793  \n",
       "  157     98.651  \n",
       "2 239     95.907  \n",
       "  218     95.760  \n",
       "  222     95.760  \n",
       "  219     95.689  \n",
       "  228     95.754  \n",
       "3 249     98.652  \n",
       "  254     97.939  \n",
       "  297     97.868  \n",
       "  300     97.868  \n",
       "  301     97.797  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "top5_blasthits_per_qid.loc[:, [\"qid\", \"Accession\", \"Scientific Name\", \"pidentity\"]].head(20)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "These don't match the hits from the publication:\n",
    "\n",
    "\n",
    "![image](strain_source_table.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Download genomes from NCBI using BioPython (Entrez)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Entrez"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "Entrez.email = \"kishored@ornl.gov\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "import urllib.request\n",
    "\n",
    "def download_file(ftp_url: str, output_file: pathlib.Path):\n",
    "    url = ftp_url.replace(\"ftp://\", \"http://\")\n",
    "    with urllib.request.urlopen(url) as url_id:\n",
    "        with open(output_file, \"wb\") as fid:\n",
    "            fid.write(url_id.read())\n",
    "\n",
    "def download_genomes(scientific_name: str, output_folder: pathlib.Path) -> None:\n",
    "    # Get list of reference or representative sequences for the given organimsm using its name\n",
    "    term = f\"({scientific_name}[ORGN]) AND (reference_genome[filter] OR representative_genome[filter])\"\n",
    "    search_handle = Entrez.esearch(db=\"assembly\", idtype=\"acc\", term=term, retmax=100)\n",
    "    search_record = Entrez.read(search_handle)\n",
    "    # Get the the accession number of the first result\n",
    "    if len(search_record[\"IdList\"]) == 0: # type: ignore\n",
    "        raise ValueError(f\"No reference or representative genomes found for {scientific_name}\")\n",
    "    if len(search_record[\"IdList\"]) > 1: # type: ignore\n",
    "        print(f\"More than one reference or representative genomes found for {scientific_name}. Using the first one.\")\n",
    "    assembly_acc = search_record[\"IdList\"][0] # type: ignore\n",
    "    # Use document summary to get the ftp path for the assembly\n",
    "    summary_handle = Entrez.esummary(db=\"assembly\", id=assembly_acc)\n",
    "    document_summary = Entrez.read(summary_handle)\n",
    "    url = document_summary['DocumentSummarySet']['DocumentSummary'][0]['FtpPath_RefSeq'] # type: ignore\n",
    "    file_id = url.split(\"/\")[-1]\n",
    "    fna_url = f\"{url}/{file_id}_genomic.fna.gz\"\n",
    "    fna_file = output_folder / f\"{file_id}.fna.gz\"\n",
    "    download_file(fna_url, fna_file)\n",
    "    gff_url = f\"{url}/{file_id}_genomic.gff.gz\"\n",
    "    gff_file = output_folder / f\"{file_id}.gff.gz\"\n",
    "    download_file(gff_url, gff_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(0, 0): Downloading fna and gff files for Scandinavium goeteborgense matching CR-75\n",
      "(0, 1): Downloading fna and gff files for Enterobacter quasihormaechei matching CR-75\n",
      "(0, 2): Downloading fna and gff files for Enterobacter hormaechei matching CR-75\n",
      "(0, 3): Downloading fna and gff files for Citrobacter freundii matching CR-75\n",
      "(0, 4): Downloading fna and gff files for Citrobacter braakii matching CR-75\n",
      "(1, 173): Downloading fna and gff files for Klebsiella aerogenes matching K-A10\n",
      "(1, 176): Downloading fna and gff files for Klebsiella aerogenes matching K-A10\n",
      "(1, 179): Downloading fna and gff files for Klebsiella aerogenes matching K-A10\n",
      "(1, 177): Downloading fna and gff files for Klebsiella aerogenes matching K-A10\n",
      "(1, 157): Downloading fna and gff files for Huaxiibacter chinensis matching K-A10\n",
      "(2, 239): Downloading fna and gff files for Pantoea endophytica matching K-A4\n",
      "(2, 218): Downloading fna and gff files for Leclercia adecarboxylata matching K-A4\n",
      "(2, 222): Downloading fna and gff files for Enterobacter sichuanensis matching K-A4\n",
      "(2, 219): Downloading fna and gff files for Leclercia adecarboxylata matching K-A4\n",
      "(2, 228): Downloading fna and gff files for Enterobacter bugandensis matching K-A4\n",
      "(3, 249): Downloading fna and gff files for Scandinavium goeteborgense matching K-A7\n",
      "(3, 254): Downloading fna and gff files for Enterobacter wuhouensis matching K-A7\n",
      "(3, 297): Downloading fna and gff files for Klebsiella aerogenes matching K-A7\n",
      "(3, 300): Downloading fna and gff files for Klebsiella aerogenes matching K-A7\n",
      "(3, 301): Downloading fna and gff files for Klebsiella aerogenes matching K-A7\n",
      "(4, 366): Downloading fna and gff files for Pseudescherichia vulneris matching K-C10\n",
      "(4, 368): Downloading fna and gff files for Pantoea endophytica matching K-C10\n",
      "(4, 364): Downloading fna and gff files for Yokenella regensburgei matching K-C10\n",
      "(4, 365): Downloading fna and gff files for Yokenella regensburgei matching K-C10\n",
      "(4, 372): Downloading fna and gff files for Klebsiella pasteurii matching K-C10\n",
      "(5, 375): Downloading fna and gff files for Kluyvera intermedia matching K-D4\n",
      "(5, 376): Downloading fna and gff files for Kluyvera intermedia matching K-D4\n",
      "(5, 377): Downloading fna and gff files for Kluyvera cryocrescens matching K-D4\n",
      "(5, 378): Downloading fna and gff files for Kluyvera cryocrescens matching K-D4\n",
      "(5, 379): Downloading fna and gff files for Kluyvera intermedia matching K-D4\n",
      "(6, 412): Downloading fna and gff files for Lelliottia amnigena matching K-D8\n",
      "(6, 401): Downloading fna and gff files for Huaxiibacter chinensis matching K-D8\n",
      "(6, 392): Downloading fna and gff files for Kluyvera intermedia matching K-D8\n",
      "(6, 420): Downloading fna and gff files for Lelliottia aquatilis matching K-D8\n",
      "(6, 404): Downloading fna and gff files for Kluyvera intermedia matching K-D8\n",
      "(7, 465): Downloading fna and gff files for Buttiauxella izardii matching K-F4\n",
      "(7, 477): Downloading fna and gff files for Buttiauxella noackiae matching K-F4\n",
      "(7, 480): Downloading fna and gff files for Lelliottia aquatilis matching K-F4\n",
      "(7, 491): Downloading fna and gff files for Pantoea endophytica matching K-F4\n",
      "(7, 462): Downloading fna and gff files for Kluyvera intermedia matching K-F4\n",
      "(8, 537): Downloading fna and gff files for Pantoea endophytica matching K-F8\n",
      "(8, 520): Downloading fna and gff files for Enterobacter sichuanensis matching K-F8\n",
      "(8, 527): Downloading fna and gff files for Enterobacter bugandensis matching K-F8\n",
      "(8, 515): Downloading fna and gff files for Enterobacter chuandaensis matching K-F8\n",
      "(8, 516): Downloading fna and gff files for Leclercia adecarboxylata matching K-F8\n",
      "(9, 547): Downloading fna and gff files for Kluyvera intermedia matching K-H10\n",
      "(9, 548): Downloading fna and gff files for Kluyvera intermedia matching K-H10\n",
      "(9, 549): Downloading fna and gff files for Kluyvera cryocrescens matching K-H10\n",
      "(9, 550): Downloading fna and gff files for Kluyvera cryocrescens matching K-H10\n",
      "(9, 551): Downloading fna and gff files for Kluyvera intermedia matching K-H10\n",
      "(10, 593): Downloading fna and gff files for Klebsiella aerogenes matching M-B5\n",
      "(10, 590): Downloading fna and gff files for Klebsiella aerogenes matching M-B5\n",
      "(10, 596): Downloading fna and gff files for Klebsiella aerogenes matching M-B5\n",
      "(10, 570): Downloading fna and gff files for Huaxiibacter chinensis matching M-B5\n",
      "(10, 594): Downloading fna and gff files for Klebsiella aerogenes matching M-B5\n",
      "(11, 612): Downloading fna and gff files for Scandinavium goeteborgense matching M-C10\n",
      "(11, 633): Downloading fna and gff files for Buttiauxella izardii matching M-C10\n",
      "(11, 638): Downloading fna and gff files for Buttiauxella agrestis matching M-C10\n",
      "(11, 625): Downloading fna and gff files for Buttiauxella massiliensis matching M-C10\n",
      "(11, 649): Downloading fna and gff files for Buttiauxella noackiae matching M-C10\n",
      "(12, 694): Downloading fna and gff files for Scandinavium goeteborgense matching M-G11\n",
      "(12, 699): Downloading fna and gff files for Enterobacter wuhouensis matching M-G11\n",
      "(12, 707): Downloading fna and gff files for Buttiauxella massiliensis matching M-G11\n",
      "(12, 705): Downloading fna and gff files for Enterobacter ludwigii matching M-G11\n",
      "(12, 697): Downloading fna and gff files for Citrobacter freundii matching M-G11\n",
      "(13, 789): Downloading fna and gff files for Pantoea agglomerans matching O-G1\n",
      "(13, 786): Downloading fna and gff files for Enterobacter ludwigii matching O-G1\n",
      "(13, 806): Downloading fna and gff files for Enterobacter sichuanensis matching O-G1\n",
      "(13, 800): Downloading fna and gff files for Leclercia adecarboxylata matching O-G1\n",
      "(13, 802): Downloading fna and gff files for Leclercia adecarboxylata matching O-G1\n",
      "(14, 101): Downloading fna and gff files for Enterobacter quasihormaechei matching escherichia-coli-k-12\n",
      "(14, 127): Downloading fna and gff files for Salmonella enterica matching escherichia-coli-k-12\n",
      "(14, 105): Downloading fna and gff files for Pseudocitrobacter vendiensis matching escherichia-coli-k-12\n",
      "(14, 116): Downloading fna and gff files for Enterobacter quasiroggenkampii matching escherichia-coli-k-12\n",
      "(14, 136): Downloading fna and gff files for Enterobacter cloacae matching escherichia-coli-k-12\n"
     ]
    }
   ],
   "source": [
    "seq_folder = pathlib.Path(\"../../data/processed/kchip/sequences/\")\n",
    "for i, row in top5_blasthits_per_qid.iterrows():\n",
    "    qid = row[\"qid\"]\n",
    "    scientific_name = \" \".join(row[\"Scientific Name\"].split(\" \")[:2])\n",
    "    print(f\"{i}: Downloading fna and gff files for {scientific_name} matching {qid}\")\n",
    "    output_folder = seq_folder / f\"{qid}\"\n",
    "    output_folder.mkdir(parents=True, exist_ok=True)\n",
    "    download_genomes(scientific_name, output_folder=output_folder)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Media information\n",
    "\n",
    "The medium used in the coculture experiment was an M9 minimal medium consisting of 1× M9 salts (Teknova), 1× trace metals (Teknova), 0.1 mM calcium chloride, and 2 mM magnesium sulfate. We additionally added 0.05% (w/v) bovine serum albumin (BSA) to the medium to improve the retention of fluorescent dyes used in the droplet color codes\n",
    "\n",
    "In the coculture experiment, a total of 40 environmental conditions were used. These included the 33 chosen compounds at 0.5% (w/v), 5 of these compounds (glucose, glycerol, pyruvate, proline, and sucrose) at 0.05% (w/v), an even mix of all 33 compounds [totaling 0.5% (w/v)], and a no-carbon control."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Microbial culture:\n",
    "\n",
    "All labeled and unlabeled monocultures initially underwent two pre-experiment regrowth cycles (“starter phase” in a rich medium and “preculture phase” in minimal medium) and, at the onset of the experiment (“experiment phase”), were normalized to a starting density of OD600 = 0.02 in carbonless minimal medium. In the starter phase, glycerol stocks of the unlabeled and labeled strains were inoculated into 525 μl (0.8-ml-deep 96-well plate) of LB medium (25°C, 220 rpm, 16 hours). Inoculations from glycerol stocks were conducted via pin replicator [sterilized via 70% (v/v) ethanol bath and flame treatment between inoculations]. In the preculture phase, all cultures were washed in carbonless M9 medium two times and then diluted (1:50) into 1-ml M9 medium with 0.5% (w/v) glucose (25°C, 220 rpm, 24 hours). Last, the experimental phase began by washing cells three times in a carbonless M9 medium to remove residual glucose and normalizing to a starting OD600 of 0.02 (or ~20 cells per droplet depending on the strain)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "kchip",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
